src/main/java/com/linecorp/armeria/server/metrics/MetricCollectingServiceCodec.java

```diff
@@ -0,0 +1,155 @@
+    private static final AttributeKey<Metrics> METRICS =
+            AttributeKey.valueOf(MetricCollectingServiceCodec.class, "METRICS");
+
+    private final ServiceMetricConsumer metricConsumer;
+
+    /**
+     * Creates a new instance that decorates the specified {@link ServiceCodec} with
+     * the specified {@link ServiceMetricConsumer}.
+     */
+    public MetricCollectingServiceCodec(ServiceCodec codec, ServiceMetricConsumer consumer) {
+        super(codec);
+        metricConsumer = consumer;
+    }
+
+    @Override
+    public DecodeResult decodeRequest(Channel ch, SessionProtocol sessionProtocol, String hostname, String path,
+                                      String mappedPath, ByteBuf in, Object originalRequest,
+                                      Promise<Object> promise) throws Exception {
+        final long startTime = System.nanoTime();
+        final int requestSize = in.readableBytes();
+
+        DecodeResult decodeResult = delegate().decodeRequest(
+                ch, sessionProtocol, hostname, path, mappedPath, in, originalRequest, promise);
+
+        LongSupplier lazyElapsedTime = () -> System.nanoTime() - startTime;
+
+        switch (decodeResult.type()) {
+        case SUCCESS: {
+            ServiceInvocationContext context = decodeResult.invocationContext();
+            context.attr(METRICS).set(new Metrics(requestSize, startTime));
+
+            promise.addListener(future -> {
+                if (!future.isSuccess()) {
+                    // encodeFailureResponse will process this case.
+                    return;
+                }
+                Object result = future.getNow();
+
+                if (result instanceof FullHttpResponse) {
+                    FullHttpResponse httpResponse = (FullHttpResponse) result;
+                    metricConsumer.invocationComplete(
+                            context.scheme(), httpResponse.status().code(),
+                            lazyElapsedTime.getAsLong(), requestSize, httpResponse.content().readableBytes(),
+                            hostname, path, mappedPath);
+                }
+                // encodeResponse will process this case.
+            });
+            break;
+        }
+        case FAILURE: {
+            final Object errorResponse = decodeResult.errorResponse();
+            if (errorResponse instanceof FullHttpResponse) {
+                FullHttpResponse httpResponse = (FullHttpResponse) errorResponse;
+                metricConsumer.invocationComplete(
+                        Scheme.of(SerializationFormat.NONE, sessionProtocol), httpResponse.status().code(),
+                        lazyElapsedTime.getAsLong(), requestSize, httpResponse.content().readableBytes(),
+                        hostname, path, mappedPath);
+            } else {
+                metricConsumer.invocationComplete(
+                        Scheme.of(SerializationFormat.NONE, sessionProtocol), HttpResponseStatus.BAD_REQUEST.code(),
```

Comment:
`SerializationFormat.NONE` is also used for the case where `SerializationFormat` does not exist really. (e.g. `SimpleHttpClient`) I guess we need to provide a dedicated enum value for 'unknown' serialization format for clarity, because otherwise a `ServiceMetricConsumer` cannot distinguish them.

Also, even if the type of the `DecodeResult` is `FAILURE` or `NOT_FOUND`, the decoder might have succeeded to determine the `SerializationFormat` of the payload. Please use `DecodeResult.decodedSerializationFormat()` instead.

If we add `SerializationFormat.UNKNOWN`, `DecodeResult.decodedSerializationFormat()` would not need to return an `Optional` anymore, which is another positive change.